File: swift/Trivia/Game.swift
//
//  Game.swift
//  Trivia
//
//  Created by Oliver Eikemeier on 13.10.15.
//  Copyright © 2015 Legacy Coderetreat. All rights reserved.
//

import Foundation

public class Game {
    var players = [String]()
    var places = [Int](count: 6, repeatedValue: 0)
    var purses  = [Int](count: 6, repeatedValue: 0)
    var inPenaltyBox  = [Bool](count: 6, repeatedValue: false)
    
    var popQuestions = [String]()
    var scienceQuestions = [String]()
    var sportsQuestions = [String]()
    var rockQuestions = [String]()
    
    var currentPlayer = 0
    var isGettingOutOfPenaltyBox = false
    
    public  init(){
    	for i in 0..<50 {
			popQuestions.append("Pop Question \(i)")
			scienceQuestions.append(("Science Question \(i)"))
			sportsQuestions.append(("Sports Question \(i)"))
			rockQuestions.append(createRockQuestion(i))
    	}
    }

    public func createRockQuestion(index: Int) -> String{
		return "Rock Question \(index)"
	}
	
	public func isPlayable() -> Bool {
		return howManyPlayers() >= 2
	}

    public func add(playerName: String) -> Bool {
		
		
	    players.append(playerName)
	    places[howManyPlayers()] = 0
	    purses[howManyPlayers()] = 0
	    inPenaltyBox[howManyPlayers()] = false
	    
	    print(playerName, "was added")
	    print("They are player number", players.count)
		return true
	}
	
	public func howManyPlayers() -> Int {
		return players.count
	}

    public func roll(roll: Int) {
		print(players[currentPlayer], "is the current player")
		print("They have rolled a", roll)
		
		if inPenaltyBox[currentPlayer] {
			if roll % 2 != 0 {
				isGettingOutOfPenaltyBox = true
				
				print(players[currentPlayer], "is getting out of the penalty box")
				places[currentPlayer] = places[currentPlayer] + roll
                if places[currentPlayer] > 11 {places[currentPlayer] = places[currentPlayer] - 12}
				
				print(players[currentPlayer]
						+ "'s new location is",
						places[currentPlayer])
				print("The category is", currentCategory())
				askQuestion()
			} else {
				print(players[currentPlayer], "is not getting out of the penalty box")
				isGettingOutOfPenaltyBox = false
				}
			
		} else {
		
			places[currentPlayer] = places[currentPlayer] + roll
            if places[currentPlayer] > 11 {places[currentPlayer] = places[currentPlayer] - 12}
			
			print(players[currentPlayer]
					+ "'s new location is",
					places[currentPlayer])
			print("The category is", currentCategory())
			askQuestion()
		}
		
	}

 	private func askQuestion() {
        if currentCategory() == "Pop" {
            print(popQuestions.removeFirst())}
        if currentCategory() == "Science"{
            print(scienceQuestions.removeFirst())}
        if currentCategory() == "Sports"{
            print(sportsQuestions.removeFirst())}
        if currentCategory() == "Rock"{
            print(rockQuestions.removeFirst())}
	}
	
	
	private func currentCategory() -> String {
        if places[currentPlayer] == 0 {return "Pop"}
		if places[currentPlayer] == 4 {return "Pop"}
		if places[currentPlayer] == 8 {return "Pop"}
		if places[currentPlayer] == 1 {return "Science"}
		if places[currentPlayer] == 5 {return "Science"}
		if places[currentPlayer] == 9 {return "Science"}
		if places[currentPlayer] == 2 {return "Sports"}
		if places[currentPlayer] == 6 {return "Sports"}
		if places[currentPlayer] == 10 {return "Sports"}
		return "Rock"
	}

	public func wasCorrectlyAnswered() -> Bool {
		if inPenaltyBox[currentPlayer]{
			if isGettingOutOfPenaltyBox {
				print("Answer was correct!!!!")
				purses[currentPlayer]++
				print(players[currentPlayer],
						"now has",
						purses[currentPlayer],
						"Gold Coins.")
				
				let winner = didPlayerWin
				currentPlayer++
                if currentPlayer == players.count {currentPlayer = 0}
				
				return winner
			} else {
				currentPlayer++
                if currentPlayer == players.count {currentPlayer = 0}
				return true
			}
			
			
			
		} else {
		
			print("Answer was corrent!!!!")
			purses[currentPlayer]++
			print(players[currentPlayer],
					"now has",
					purses[currentPlayer],
					"Gold Coins.")
			
			let winner = didPlayerWin
			currentPlayer++
            if currentPlayer == players.count {currentPlayer = 0}
			
			return winner
		}
	}
	
	public func wrongAnswer()->Bool{
		print("Question was incorrectly answered")
		print(players[currentPlayer], "was sent to the penalty box")
		inPenaltyBox[currentPlayer] = true
		
		currentPlayer++
        if currentPlayer == players.count {currentPlayer = 0}
		return true
	}


    private var didPlayerWin: Bool {
		return !(purses[currentPlayer] == 6)
	}
}

File: clojure/src/trivia/Game.clj
(ns trivia.Game)

(def players (atom []))
(def places (atom [0] ))
(def purses (atom [0]))
(def inPenaltyBox (atom [false]))
(def popQuestions (atom []))
(def scienceQuestions (atom []))
(def sportsQuestions (atom []))
(def rockQuestions (atom []))
(def currentPlayer (atom 0))
(def isGettingOutOfPenaltyBox (atom false))

(defn createRockQuestion [index] (str "Rock Question " index))

(defn initialize []
  (dotimes [i 50]
    (swap! popQuestions conj (str "Pop Question " i))
    (swap! scienceQuestions conj (str "Science Question " i))
    (swap! sportsQuestions conj (str "Sports Question " i))
    (swap! rockQuestions conj (createRockQuestion i))))

(initialize)

(defn howManyPlayers [] (count @players))

(defn isPlayable [] (>= (howManyPlayers) 2))

(defn add [playerName]
  (swap! players conj playerName)
  (swap! places assoc (howManyPlayers) 0)
  (swap! purses assoc (howManyPlayers) 0)
  (swap! inPenaltyBox assoc (howManyPlayers) false)
  (println playerName "was added")
  (println "They are player number " (count @players)))

(defn currentCategory []
  (cond (= (nth @places @currentPlayer) 0) "Pop"
        (= (nth @places @currentPlayer) 4) "Pop"
        (= (nth @places @currentPlayer) 8) "Pop"
        (= (nth @places @currentPlayer) 1) "Science"
        (= (nth @places @currentPlayer) 5) "Science"
        (= (nth @places @currentPlayer) 9) "Science"
        (= (nth @places @currentPlayer) 2) "Sports"
        (= (nth @places @currentPlayer) 6) "Sports"
        (= (nth @places @currentPlayer) 10) "Sports"
        :else "Rock"))

(defn askQuestion []
  (cond (= (currentCategory) "Pop") (do (println (first @popQuestions)) (swap! popQuestions #(drop 1 %)))
        (= (currentCategory) "Science") (do (println (first @scienceQuestions)) (swap! scienceQuestions #(drop 1 %)))
        (= (currentCategory) "Sports") (do (println (first @sportsQuestions)) (swap! sportsQuestions #(drop 1 %)))
        (= (currentCategory) "Rock") (do (println (first @rockQuestions)) (swap! rockQuestions #(drop 1 %)))))

(defn roll [roll]
  (println (nth @players @currentPlayer) "is the current player")
  (println "They have rolled a" roll)
  (if (nth @inPenaltyBox @currentPlayer)
    (if (not= 0 (mod roll 2))
      (do (reset! isGettingOutOfPenaltyBox true)
          (println (nth @players @currentPlayer) "is getting out of the penalty box")
          (swap! places assoc @currentPlayer (+ (@places @currentPlayer) roll))
          (if (> (nth @places @currentPlayer) 11)
            (swap! places assoc @currentPlayer (- (nth @places @currentPlayer) 12)))
          (println (str (nth @players @currentPlayer) "'s new location is") (nth @places @currentPlayer))
          (println "The category is" (currentCategory))
          (askQuestion))
      (do (println (nth @players @currentPlayer) "is not getting out of the penalty box")
          (reset! isGettingOutOfPenaltyBox false)))
    (do (swap! places assoc @currentPlayer (+ (@places @currentPlayer) roll))
        (if (> (nth @places @currentPlayer) 11)
          (swap! places assoc @currentPlayer (- (nth @places @currentPlayer) 12)))
        (println (str (nth @players @currentPlayer) "'s new location is") (nth @places @currentPlayer))
        (println "The category is" (currentCategory))
        (askQuestion))))

(defn didPlayerWin []
  (not (= (nth @purses @currentPlayer) 6)))

(defn wasCorrectlyAnswered []
  (if (nth @inPenaltyBox @currentPlayer)
    (if @isGettingOutOfPenaltyBox
      (do (println "Answer was correct!!!!")
          (swap! purses assoc @currentPlayer (inc (nth @purses @currentPlayer)))
          (println (nth @players @currentPlayer) "now has" (nth @purses @currentPlayer) "Gold Coins.")
          (let [winner (didPlayerWin)]
            (swap! currentPlayer inc)
            (if (= @currentPlayer (count @players)) (reset! currentPlayer 0))
            winner))
      (do (swap! currentPlayer inc)
          (if (= @currentPlayer (count @players)) (reset! currentPlayer 0))
          true))
    (do (println "Answer was corrent!!!!")
        (swap! purses assoc @currentPlayer (inc (nth @purses @currentPlayer)))
        (println (nth @players @currentPlayer) "now has" (nth @purses @currentPlayer) "Gold Coins.")
          (let [winner (didPlayerWin)]
            (swap! currentPlayer inc)
            (if (= @currentPlayer (count @players)) (reset! currentPlayer 0))
            winner))))

(defn wrongAnswer []
  (println "Question was incorrectly answered")
  (println (nth @players @currentPlayer) "was sent to the penalty box")
  (swap! inPenaltyBox assoc @currentPlayer true)
  (swap! currentPlayer inc)
  (if (= @currentPlayer (count @players)) (reset! currentPlayer 0))
  true)
